sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/StringSqlAggregator.java

```diff
@@ -0,0 +1,219 @@
+        return null;
+      }
+      maxSizeBytes = ((Number) RexLiteral.value(maxBytes)).intValue();
+    }
+    final DruidExpression arg = arguments.get(0);
+    final ExprMacroTable macroTable = plannerContext.getExprMacroTable();
+
+    final String initialvalue = "[]";
+    final ValueType elementType = ValueType.STRING;
+    final String fieldName;
+    if (arg.isDirectColumnAccess()) {
+      fieldName = arg.getDirectColumn();
+    } else {
+      VirtualColumn vc = virtualColumnRegistry.getOrCreateVirtualColumnForExpression(plannerContext, arg, elementType);
+      fieldName = vc.getOutputName();
+    }
+
+    final String finalizer = StringUtils.format("if(array_length(o) == 0, null, array_to_string(o, '%s'))", separator);
+    final NotDimFilter dimFilter = new NotDimFilter(new SelectorDimFilter(fieldName, null, null));
+    if (aggregateCall.isDistinct()) {
+      return Aggregation.create(
+          // string_agg ignores nulls
+          new FilteredAggregatorFactory(
+              new ExpressionLambdaAggregatorFactory(
+                  name,
+                  ImmutableSet.of(fieldName),
+                  null,
+                  initialvalue,
+                  null,
+                  StringUtils.format("array_set_add(\"__acc\", \"%s\")", fieldName),
+                  StringUtils.format("array_set_add_all(\"__acc\", \"%s\")", name),
+                  null,
+                  finalizer,
+                  maxSizeBytes != null ? new HumanReadableBytes(maxSizeBytes) : null,
+                  macroTable
+              ),
+              dimFilter
+          )
+      );
+    } else {
+      return Aggregation.create(
+          // string_agg ignores nulls
+          new FilteredAggregatorFactory(
+              new ExpressionLambdaAggregatorFactory(
+                  name,
+                  ImmutableSet.of(fieldName),
+                  null,
+                  initialvalue,
+                  null,
+                  StringUtils.format("array_append(\"__acc\", \"%s\")", fieldName),
+                  StringUtils.format("array_concat(\"__acc\", \"%s\")", name),
+                  null,
+                  finalizer,
+                  maxSizeBytes != null ? new HumanReadableBytes(maxSizeBytes) : null,
+                  macroTable
+              ),
+              dimFilter
+          )
+      );
+    }
+  }
+
+  private static class StringAggFunction extends SqlAggFunction
+  {
+    StringAggFunction()
+    {
+      super(
+          NAME,
+          null,
+          SqlKind.OTHER_FUNCTION,
+          opBinding ->
+              Calcites.createSqlTypeWithNullability(opBinding.getTypeFactory(), SqlTypeName.VARCHAR, true),
+          InferTypes.ANY_NULLABLE,
+          OperandTypes.or(
+              OperandTypes.and(
+                  OperandTypes.sequence(
+                      StringUtils.format("'%s'(expr, separator)", NAME),
+                      OperandTypes.ANY,
+                      OperandTypes.LITERAL
```

Comment:
I'm not certain if this can (or should) handle an expression for the `separator` parameter, I'll do some investigation and potentially update this signature accordingly

Reply:
btw, I'm going to hold off on updating this PR until #11280 goes in, since it has a slight refactor/improvement of the underlying expression aggregator that will require some minor changes in this PR.